### PR TITLE
feat(evals): build and publish trace-derived evals locally

### PR DESCRIPTION
**Pull request by aamir**
## Summary

A developer can now ask their coding agent to turn a hosted Understudy workload into a checked, publishable eval without using the web app. The CLI downloads and authenticates the complete frozen week, keeps workload reasoning local, proves the verifier against independent good and wrong fixtures, and uploads only the exact owner-approved release.

This is the agent-tools half of the CLI-first eval workflow. It uses the companion platform PR for complete trace export and immutable release storage. Hosted model runs, GEPA, and cleanup of the old web authoring path remain separate follow-up work.

Related: understudylabs/understudy-platform#646

## Workflow

1. `capture-evidence` runs `understudy evals build --last 7d` and materializes every receipt- and digest-verified export segment into one private local project.
2. The coding agent searches the compact execution index, reconstructs provable lineage, confirms the workload profile and metric with the owner, and authors representative tasks, coverage, environment, and verifier.
3. `understudy evals check` replays immutable module snapshots in fresh deterministic workers. It requires a known-good pass, an intentionally wrong rejection, repeatability, complete source accounting, and bound owner approvals.
4. `understudy evals publish --preview` prints the exact manifest, release ID, bundle hash, size, and ordered inventory without a network request.
5. Upload requires a matching `--expect-release-id`. Any local change invalidates the preview before the CLI sends data.

## Data and execution boundaries

- The whole week is stored locally, but the agent can inspect the compact index and individual traces instead of placing the corpus in one prompt.
- Repository builds must use a child directory under `.understudy`; staging and downloaded payloads remain inside that ignored private tree, and an empty authenticated export fails before a project is created.
- The backend freezes ingestion at export start, at or after the exact seven-day trace window. The CLI omits a cutoff on page one, preserves the returned canonical cutoff across committed resumed segments, and safely restarts an interrupted first page so already-arrived, late-ingested rows from the week are included without caller-selected or later corpus drift.
- Each local trace must match the content SHA-256 in the authenticated export manifest. Fresh downloads and resumed same-size files fail closed on a digest mismatch.
- The backend and CLI independently derive the same ordered rolling commitment from every source row's identity, byte size, and content digest. The verified terminal export binds that exact corpus into a compact durable source attestation.
- The check report and publication hash the exact attestation token, and the backend verifies it against the frozen workload scope, totals, and corpus commitment before sealing a release.
- Publication sends exactly two reviewed objects: the displayed manifest, which contains the attestation and its hash, and its gzip bundle; the ordered inventory rejects file/descendant path overlaps so the archive can always be materialized exactly.
- The bundle excludes `source/`, raw traces, the expiring receipt and local export proof, the execution index, analysis, and mutable `eval-project.json` state.
- The historical incumbent answer is never treated as gold. Both positive and negative fixtures require independent correctness evidence.
- Build, check, and publish do not call a model, start a hosted eval, change a prompt, or mutate serving.
- The local JavaScript worker is a bounded deterministic verifier runtime, not an operating-system sandbox for arbitrary untrusted code.

Session-settled decisions carried from planning: the coding agent owns eval semantics; the source is a complete frozen week; the backend only transports evidence and stores the final release; publication is a separate explicit action.

## Validation

- `npm run check`: 1,275 tests passed across 165 suites.
- Public skill validation: 43 skills passed.
- Package dry-run smoke passed.
- Packaged JSON Schemas name the exported runtime validators required for cross-field release invariants; platform golden publication/release bytes match.
- Review covered correctness, upload consent, local data boundaries, deterministic execution, API parity, reliability, tests, public-repo standards, and agent usability. Every actionable finding to date was fixed and its thread resolved.

## Deliberately deferred

- Running models or prompts against the release, background benchmarks, GEPA, and result-ledger integration.
- Treating the local module worker as a general-purpose hostile-code sandbox.
- Removing the existing web eval authoring code before Cedar validates this flow.

## Post-Deploy Monitoring & Validation

- Owner/window: an Understudy operator stays with Cedar through the first end-to-end run and reviews the first published release before any later benchmark work.
- Start with a synthetic or Understudy-owned workload. Verify `evals build` reaches a terminal seven-day receipt and that every local source count, byte total, and content digest matches the authenticated export.
- Verify `evals check` reports complete lineage accounting, explicit coverage gaps, one good pass, one wrong rejection, stable repeated results, and the final artifact hashes.
- Review the complete `evals publish --preview` inventory. Upload only after explicit permission and only with its exact `--expect-release-id`.
- Healthy publication returns the expected release ID; platform readback must return the same manifest and bundle hashes. A changed preview, source-accounting mismatch, unexpected upload file, `4xx` scope/approval error, or `5xx` is a stop signal.
- Rollback: revert the package change or keep Cedar on the previous CLI version. No eval command in this PR can modify live routing, prompts, models, or serving traffic.

---

[Compound Engineering](https://github.com/EveryInc/compound-engineering-plugin)

---

<a href="https://app.devin.ai/review/understudylabs/understudy-agent-tools/pull/476" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-devin-review-dark.svg?v=3">
    Devin Review
  </picture>
</a>

### CODE REVIEW

File: src/eval-project.ts [hunk 1]
```diff
@@ -137,3 +216,99 @@ export function buildEvalProject(options: BuildEvalProjectOptions): EvalProjectB
...
+      window: options.canonicalScope,
+      capture_count: files.length,
+      size_bytes: files.reduce((sum, file) => sum + file.size_bytes, 0),
+      index: portableRelative(projectRoot, indexPath),
+      index_sha256: indexSha256,
+      export_proof: portableRelative(projectRoot, proofPath),
+      export_proof_sha256: createHash("sha256").update(proofBody).digest("hex"),
+      exported_capture_count: options.verifiedReceipt.cumulative_exported,
+      exported_total_bytes: options.verifiedReceipt.total_bytes,
+      terminal_receipt_verified: true,
```
[the remark this answers is not in the document]
Comment (aamir) [reply src/eval-project.ts:290]: > When export pages repeat a capture, `capture_count` uses unique files while receipt totals include repeats.

Addressed in `7be6ecef`: project construction now reconciles the deduplicated capture count and byte total against the verified terminal receipt before writing the source index or eval manifest. A repeated key across segments fails before either final artifact is published.

File: src/evals/check.ts [hunk 2]
```diff
@@ -0,0 +1,654 @@
...
+    coverage_sha256: coverageSha256,
+    environment_sha256: environmentSha256,
+    verifier_sha256: verifierSha256,
+  });
+  const checkReportPath = reportPath(projectRoot, project.artifacts.check_report);
+  let report = candidateReport;
+  try {
+    const existing = parseJson(checkReportPath, EvalCheckReportSchema, "checks/report.json");
+    if (sameReport(existing, candidateReport)) report = existing;
+    else replacePrivateJson(checkReportPath, candidateReport);
```
[the remark this answers is not in the document]
Comment (aamir) [reply src/evals/check.ts]: > After a clock rollback, `sameReport` reuses a report dated after the current check.

Addressed in `7be6ecef`: a semantically identical report is reused only when its timestamp is not later than the current check time. The chronology regression rewinds the clock, regenerates the report, and proves a later owner approval can publish.

File: src/evals/materialize.ts [hunk 3]
```diff
@@ -183,6 +191,231 @@ export function reserveDownloadedChunk(
...
+    }
+    if (sizeBytes !== item.size) {
+      throw new Error(`Capture ${item.request_id} ended before its authenticated manifest size.`);
+    }
+    closeSync(descriptor);
+    descriptor = null;
+    const digest = hash.digest("hex");
+    renameSync(partialPath, finalPath);
+    complete = true;
+    return { digest, sizeBytes };
```
[the remark this answers is not in the document]
Comment (aamir) [reply src/evals/materialize.ts:394]: > Include a content SHA-256 in the authenticated manifest and reject the download unless the computed digest matches it.

Addressed in `7be6ecef`: every workload export manifest item now requires `content_sha256`, and materialization checks it for fresh downloads, checkpointed files, and same-size untracked recovery files before publishing local evidence. A focused corruption regression covers both download and recovery paths; the paired platform contract is being updated to emit the digest in its authenticated manifest.

File: src/evals/authoring-contracts.ts [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+  VerifiedWorkloadCaptureFileSchema,
+  VerifyWorkloadCaptureExportReceiptResponseSchema,
+  WorkloadCaptureExportScopeSchema,
+} from "./contracts.js";
+
+const TimestampSchema = z.string().datetime();
+const RelativeArtifactPathSchema = z.string().min(1).refine(
+  (value) => !value.startsWith("/") && !value.includes("\\") && !/^[A-Za-z]:[\\/]/.test(value) && !value.split("/").includes(".."),
+  "artifact paths must be project-relative and cannot contain '..'",
+);
```
[the remark this answers is not in the document]
Comment (aamir) [reply src/evals/authoring-contracts.ts]: > Paths with dot segments, empty segments, or over 240 characters pass `RelativeArtifactPathSchema`.

Addressed in `7be6ecef`: local authoring, checker path resolution, and release publication now share the exact normalized 240-character path contract. The new regression covers invalid components and a real check-to-publish flow with a portable nested artifact path.

File: schemas/understudy.eval-publication.v1.schema.json [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    "workload_id": { "$ref": "#/$defs/id" },
+    "eval_id": { "type": "string", "pattern": "^eval_[a-f0-9]{24}$" },
+    "name": { "type": "string", "minLength": 1, "maxLength": 120 },
+    "source": { "$ref": "#/$defs/source" },
+    "artifacts": { "$ref": "#/$defs/artifacts" },
+    "runtime": { "$ref": "#/$defs/runtime" },
+    "skills": { "type": "array", "minItems": 1, "maxItems": 32, "items": { "$ref": "#/$defs/skill" } },
+    "approval": { "$ref": "#/$defs/approval" },
+    "artifact_layout": { "$ref": "#/$defs/layout" },
+    "bundle_files": { "type": "array", "minItems": 1, "maxItems": 1024, "items": { "$ref": "#/$defs/bundle_file" } }
```
[the remark this answers is not in the document]
Comment (aamir) [reply schemas/understudy.eval-publication.v1.schema.json]: > Duplicate, unsorted, or undeclared bundle paths pass the public `bundle_files` schema.

Addressed in `7be6ecef` using the suggested shared-validation-contract option: the packaged JSON Schemas explicitly identify their structural-only boundary and the required package-exported semantic validator. `EvalPublicationSchema` and `EvalReleaseSchema` are now public exports, exact duplicate items also fail structurally, and drift tests cover duplicate, unsorted, missing, undeclared, overlapping-root, and misplaced-entrypoint inventories.

**Comment by aamir** [comment]
@codex review

**Comment by aamir** [comment]
@codex review

**Comment by aamir** [comment]
@codex review

File: src/commands/traces.ts [hunk 6]
```diff
@@ -33,8 +33,13 @@ export function registerTracesCommand(program: Command): void {
...
     .option("--workload <name>", "Compile only captures matching workload id or name")
     .option("--batch-size <count>", "Resumable processing batch size", "10")
-    .action((options: { source: string; output: string; maxAgeDays: string; workload?: string; batchSize: string }) => {
-      const result = compileTraceFoundry(resolve(options.source), resolve(options.output), Number(options.maxAgeDays), new Date(), { workload: options.workload, batchSize: Number(options.batchSize) });
+    .option("--reference-time <iso>", "Reference time for freshness (hosted evals use eval-project.json source.window.to)")
+    .option("--provable-lineage-only", "Exclude ambiguous and unlinked executions from generated tasks")
+    .option("--source-index <path>", "Frozen eval source/index.jsonl that binds every hosted capture file")
+    .action((options: { source: string; output: string; maxAgeDays: string; workload?: string; batchSize: string; referenceTime?: string; provableLineageOnly?: boolean; sourceIndex?: string }) => {
+      const referenceTime = options.referenceTime === undefined ? new Date() : new Date(options.referenceTime);
```
[the remark this answers is not in the document]
Comment (aamir) [reply src/commands/traces.ts]: Accepted and fixed in 3ec1c437. `--provable-lineage-only` now requires an explicit `--reference-time` from the frozen eval window; ordinary local compilation keeps its wall-clock default. Added positive/negative coverage, and the full `npm run check` passes 1,274/1,274 tests.

**Comment by aamir** [comment]
@codex review

File: src/commands/evals.ts [hunk 7]
```diff
@@ -269,123 +358,264 @@ async function runBuild(cmd: Command, opts: BuildOpts): Promise<void> {
...
   opts: BuildOpts,
-  build: { batchSize: number; windowMs: number; maxAgeDays: number; output: string; selection: EvalBuildSelection },
+  build: { batchSize: number; windowMs: number; maxAgeDays: number; output: string },
 ): Promise<void> {
-  const { batchSize, windowMs, maxAgeDays, output, selection } = build;
+  const { batchSize, windowMs, maxAgeDays, output } = build;
   if (pathExists(output)) {
     throw new Error(`Eval build destination already exists: ${output}. Choose a fresh --out directory.`);
   }
   const staging = join(dirname(output), `.${basename(output)}.eval-build`);
```
[the remark this answers is not in the document]
Comment (aamir) [reply src/commands/evals.ts:387]: Accepted and fixed in 0ecd4c22. The exact `.understudy` repository root is now reserved: builds must use a child directory, and the guard runs before any hosted read or staging lease. The CLI regression proves no `..understudy.eval-build` sibling is created.

File: src/commands/evals.ts [hunk 8]
```diff
@@ -269,123 +358,264 @@ async function runBuild(cmd: Command, opts: BuildOpts): Promise<void> {
...
+  const receipt = await verifyWorkloadExportReceipt(context, state);
+  if (
+    receipt.chain_id !== state.transport.chain_id ||
+    receipt.cumulative_exported !== state.transport.cumulative_exported ||
+    receipt.total_bytes !== state.transport.cumulative_total_bytes ||
+    receipt.manifest_sha256 !== state.transport.previous_manifest_sha256 ||
+    receipt.local_index_sha256 !== sourceIndexCommitmentSha256(state.transport.verified_files)
+  ) throw new Error("Verified capture export receipt does not match the downloaded source chain.");
+
+  const project = buildWorkloadEvalProject({
```
[the remark this answers is not in the document]
Comment (aamir) [reply src/commands/evals.ts:488]: Accepted and fixed in 0ecd4c22. A verified terminal export with zero captures is now rejected before any eval project directory is created. Added a direct regression, and the full `npm run check` passes 1,275/1,275 tests.

**Comment by aamir** [comment]
@codex review

**Comment by aamir** [comment]
@codex review
